sdk/keyvault/azure-security-keyvault-keys/src/key_client.cpp

```diff
@@ -187,6 +192,12 @@ Azure::Response<KeyPropertiesSinglePage> KeyClient::GetPropertiesOfKeyVersionsSi
       },
       request.Path,
       request.Query);
+  KeyPropertiesSinglePage responsePage = std::move(response.Value);
```

Comment:
This is a strange way to construct the type.  This should be a constructor.

Reply:
FYI, this is similar to the pattern that storage is currently using to create the paged response types in the client libraries.

The implementation detail pattern and design of ctors on derived types needs to be fleshed out.
This issue is related to it: https://github.com/Azure/azure-sdk-for-cpp/issues/2152

@JeffreyRichter We need to provide guidance arounds this. What do you recommend?

For example:
https://github.com/Azure/azure-sdk-for-cpp/blob/698e9965c56c47cf16d9350956e8d752031e5fe5/sdk/storage/azure-storage-blobs/src/blob_service_client.cpp#L134-L144

Reply:
Thinks to consider:
- A new cttor will be required to be added every time there is a new field added to the model T.
- The cttors are created based on the needs and implementation of an external component. For example, if `A` needs to create `B`, should we make `B` cttors based on the specific needs of `A`?   I would think `A` and `B` should be considered as individual components which design/cttors are not based on other's needs. The design should be created based on how we want that component to behave/work

My personal way to use cttors is:
- To set the init - private state of the component. Just for the encapsulated stated / private fields.  
- Do not init public fields, as they are not relevant, as per component design, for the initial state and they can be set/updated/etc at any point.

Reply:
This is a tough one because the language has no way of exposing an "internal" member (like C# does).
So, it comes down to least evil thing to do. Do you prefer a "public" ctor that is only called by SDK code and is therefore documented as "DO NOT USE" for customers? Or, do you prefer to violate encapsulation via a friend?
Or, maybe @CaseyCarter or @BillyONeal have another suggestion for to better handle this.

Since I personally value the customer experience as paramount, I guess I lean towards the friend "solution".

Reply:
> My personal way to use cttors is:
> 
> * To set the init - private state of the component. Just for the encapsulated stated / private fields.
> * Do not init public fields, as they are not relevant, as per component design, for the initial state and they can be set/updated/etc at any point.

+1 - I like the distinction between private state and public fields. The ctor must manage the init of private fields, and those shouldn't be set outside of the class. It _can_ also set the public fields, but isn't required.

Reply:
I'd vote for friend method, as it's less visible to customers.